angr/state_plugins/concrete.py

```diff
@@ -0,0 +1,137 @@
+
+                    func_address = target.read_memory(reloc.rebased_addr, self.state.project.arch.bits / 8)
+                    func_address = struct.unpack(self.state.project.arch.struct_fmt(), func_address)[0]
+                    self.state.project.rehook_symbol(func_address, reloc.symbol.name)
+        else:
+            l.warn("SimProc not restored, you are going to simulate also the code of external libraries!")
+
+        # flush the angr memory in order to synchronize them with the content of the
+        # concrete process memory when a read/write to the page is performed
+        self.state.memory.flush_pages(self.whitelist)
+        l.info("Exiting SimEngineConcrete: simulated address %x concrete address %x "
+               % (self.state.addr, target.read_register("pc")))
+
+        # now we have to register a SimInspect in order to synchronize the segments register
+        # on demand when the symbolic execution accesses it
+        if not self.segment_registers_callback_initialized:
+            self.fs_register_bp = self.state.inspect.b('reg_read', reg_read_offset=self.state.project.simos.get_segment_register_name(),
+                                                       action=self._sync_segments)
+
+            self.segment_registers_callback_initialized = True
+
+            l.debug("Set SimInspect breakpoint to the new state!")
+
+    def _sync_registers(self, register_names, target):
+        for register_name in register_names:
+            try:
+                reg_value = target.read_register(register_name)
+                setattr(self.state.regs, register_name, reg_value)
+                l.debug("Register: %s value: %x " % (register_name,
+                                                     self.state.se.eval(getattr(self.state.regs, register_name),
+                                                                        cast_to=int)))
+            except ConcreteRegisterError as exc:
+                l.debug("Can't set register %s reason: %s, if this register is not used "
+                        "this message can be ignored" % (register_name, exc))
+
+    def _sync_segments(self, state):
+        """
+        Segment registers synchronization is on demand as soon as the
+        symbolic execution access a segment register.
+        """
+        target = state.project.concrete_target
+
+        if isinstance(state.arch, ArchAMD64):
+            state.project.simos.initialize_segment_register_x64(state, target)
+        elif isinstance(state.arch, ArchX86):
+            gdt = state.project.simos.initialize_gdt_x86(state, target)
+            state.concrete.whitelist.append((gdt.addr, gdt.addr + gdt.limit))
+
+        state.inspect.remove_breakpoint('reg_read', bp=state.concrete.fs_register_bp)
+        state.concrete.segment_registers_initialized = True
+
+        state.concrete.fs_register_bp = None
+
+
+from angr.sim_state import SimState
```

Comment:
Relative import please.

Reply:
I noticed that all the SimStatePlugin use this absolute import should I keep it like the others or use the relative import? Should I refactor also the other occurences?

Reply:
We should've caught them earlier. Please only use relative imports unless you are writing `SimProcedure`s.